psasim/test/server.c

```diff
@@ -47,6 +47,11 @@ int psa_sha256_main(int argc, char *argv[])
     parse_input_args(argc, argv);
     SERVER_PRINT("Starting");
 
+    int pid = fork();
+    if (pid != 0) {
+        exit(0);
+    }
```

Comment:
I think we need `setsid()` in the child of fork (i.e. after line 53) as part of standard demonisation. Demonisation also usually redirects stdin, stdout and stderr to `/dev/null`, but we may not want to do that.

It might be useful to have a way to keep the server from demonising in case someone does want to run it at a terminal for debugging (this "someone" is likely to be us in a few weeks when we're trying to track down a bug somewhere!)

Reply:
Given that this is purely a testing tool, why would we even want daemonization? We expect the daemon's lifetime to be controlled by a test script, not by a service monitor.

Reply:
@tom-cosgrove-arm suggested that instead of calling the server with the final "&" to put it into the background, we can actually daemonize it (by using a specific command line argument). Based on your comment it seems that this is a bit overengineered, ins't it?

Reply:
When we're testing, we can often expect the server or the whole test to misbehave. So killing the server should be easy, and killing the process group should not leave a stray process. Daemonization breaks those two properties.

When we're not testing, daemonization is desirable. But when would we ever use the psasim server for purposes other than testing?

Reply:
There are two discrete use cases. One is the user who is testing locally and wants the script to exit as @valeriosetti  has done, and the other is the CI which will more likely will want the server to block and hit it with clients. So I also see @gilles-peskine-arm  point.

How about a compromise?
- We restore the original functionality to block
- We add a configuration option or parameter that make it exit when client disconnects for local user testing?

If a deamon is needed we can extend those optional parameters in the future?

Reply:
When I mentioned testing above, I was thinking both of ad hoc runs during development, and CI scripts. For ad hoc runs, I want to run the server, perhaps in a debugger, and kill it with Ctrl+C. For CI scripts, I want to run the server, and `kill(1)` or `kill(2)` it when done.

> We add a configuration option or parameter that make it exit when client disconnects for local user testing?

Hmmm, maybe. We'll have some tests where we want to exercise persistence between client connections, but most tests won't do that. On the other hand, when running a test suite, I think we'll want to run the server once for the whole suite (to avoid a relatively large overhead per test case, and to avoid having to code up something to make the client autostart the server), but have the client connect separately in each test case (because each test case calls `psa_crypto_init` and `mbedtls_psa_crypto_free`). So I don't see a lot of use for an auto-exit on disconnect.

Reply:
Ok, so if I understood correctly I just need to remove the last commit (going back to https://github.com/Mbed-TLS/mbedtls-framework/pull/10/commits/e91126b960e6b7e4c485d1a83266a58cc986ea88) and there I just need to invert the logic meaning of "k" parameter. Now the default behavior will be keep server alive if no "k" param is defined, otherwise "kill on client disconnection".